From c02b8506469203e537fdf6fe37d4447cc4d358a3 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Wed, 1 Apr 2026 16:37:01 +0100
Subject: [PATCH] cow: Support overlays larger than 16T on ext4

nbdkit-cow-filter writes the copy-on-write overlay into a temporary
file under /var/tmp which has the same size as the underlying virtual
disk.  If the temporary directory being used is on ext4, this
effectively limits the size of the overlay to just under 16T, since
that is the maximum file size supported.  (Note that the file is
sparse so this does not mean all the space is used.).  For example:

  $ nbdkit --filter=cow null 17T --run 'nbdinfo "$uri"'
  nbdkit: null[1]: error: ftruncate: File too large
  nbdkit: null[1]: error: ftruncate: File too large
  nbdkit: null[1]: error: ftruncate: File too large
  nbdinfo: nbd_opt_go: the server has no export named '': ftruncate: File too large: No such file or directory for the default export
  nbdinfo: suggestion: to list all exports on the server, use --list
  protocol: newstyle-fixed without TLS, using structured packets

This change splits the overlay into as many 8T chunks as needed, so
the ext4 maximum file size is no longer a limit.

Fixes: https://issues.redhat.com/browse/RHEL-163983
(cherry picked from commit eb79a0e0c63ba1884bde02dc884a26aea2fc4324)
---
 filters/cow/blk.c      | 69 ++++++++++++++++++++++++++++++++--
 tests/Makefile.am      |  6 ++-
 tests/test-cow-huge.sh | 85 ++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 155 insertions(+), 5 deletions(-)
 create mode 100755 tests/test-cow-huge.sh

diff --git a/filters/cow/blk.c b/filters/cow/blk.c
index 1012fb5e..defd0d94 100644
--- a/filters/cow/blk.c
+++ b/filters/cow/blk.c
@@ -67,6 +67,10 @@
  *
  * Since the overlay is a deleted temporary file, we can ignore FUA
  * and flush commands.
+ *
+ * For disk sizes larger than 8T, the temporary file is split into
+ * multiple files, at most 8T in size.  This ensures we can work on
+ * filesystems with file size limits like ext4 (max file size < 16T).
  */
 
 #include <config.h>
@@ -81,6 +85,7 @@
 #include <fcntl.h>
 #include <limits.h>
 #include <errno.h>
+#include <assert.h>
 #include <sys/types.h>
 
 #ifdef HAVE_ALLOCA_H
@@ -109,6 +114,8 @@ enum bm_entry {
   BLOCK_TRIMMED = 3,
 };
 
+#define MAX_FILE_SIZE (UINT64_C (8) * 1024 * 1024 * 1024 * 1024) /* 8T */
+
 static const char *
 state_to_string (enum bm_entry state)
 {
@@ -256,13 +263,56 @@ int
 blk_set_size (struct blk_overlay *blk, uint64_t new_size)
 {
   ACQUIRE_LOCK_FOR_CURRENT_SCOPE (&blk->lock);
+  const uint64_t new_size_rounded_up = ROUND_UP (new_size, blksize);
+  uint64_t remainder;
+  size_t i, n;
+  int fd;
 
   blk->size = new_size;
 
   if (bitmap_resize (&blk->bm, blk->size) == -1)
     return -1;
 
-  if (ftruncate (blk->fds.ptr[0], ROUND_UP (blk->size, blksize)) == -1) {
+  /* How many temporary files do we need?  Each temp file is limited to 8T. */
+  n = DIV_ROUND_UP (new_size_rounded_up, MAX_FILE_SIZE);
+  if (n == 0)
+    n = 1;
+  nbdkit_debug ("cow: using %zu temporary file(s)", n);
+
+  /* Create new temporary files in case it has expanded. */
+  while (blk->fds.len < n) {
+    fd = create_temporary_file ();
+    if (fd == -1)
+      return -1;
+    if (fd_vector_append (&blk->fds, fd) == -1) {
+      nbdkit_error ("realloc: %m");
+      /* The other fds are stored in the handle and will be closed by
+       * blk_free, so we only need to close this one.
+       */
+      close (fd);
+      return -1;
+    }
+  }
+
+  /* Remove any extra temporary files in case it has shrunk. */
+  while (blk->fds.len > n) {
+    close (blk->fds.ptr[n]);
+    fd_vector_remove (&blk->fds, n);
+  }
+
+  /* The first 0..n-1 temp files have to be truncated to the full
+   * size, with the last temp file being truncated to the remainder.
+   */
+  remainder = new_size_rounded_up;
+  for (i = 0; i < n-1; ++i) {
+    if (ftruncate (blk->fds.ptr[i], MAX_FILE_SIZE) == -1) {
+      nbdkit_error ("cow: ftruncate: %m");
+      return -1;
+    }
+    assert (remainder >= MAX_FILE_SIZE);
+    remainder -= MAX_FILE_SIZE;
+  }
+  if (ftruncate (blk->fds.ptr[n-1], remainder) == -1) {
     nbdkit_error ("cow: ftruncate: %m");
     return -1;
   }
@@ -270,6 +320,15 @@ blk_set_size (struct blk_overlay *blk, uint64_t new_size)
   return 0;
 }
 
+/* Get the temporary file corresponding to a block. */
+static int
+get_fd_for_blknum (struct blk_overlay *blk, uint64_t blknum)
+{
+  size_t i = blknum / MAX_FILE_SIZE / blksize;
+  assert (i < blk->fds.len);
+  return blk->fds.ptr[i];
+}
+
 /* This is a bit of a hack since usually this information is hidden in
  * the blk module.  However it is needed when calculating extents.
  */
@@ -296,7 +355,7 @@ blk_read_multiple (struct blk_overlay *blk,
   off_t offset = blknum * blksize;
   enum bm_entry state;
   uint64_t b, runblocks;
-  const int fd = blk->fds.ptr[0];
+  const int fd = get_fd_for_blknum (blk, blknum);
 
   /* Find out how many of the following blocks form a "run" with the
    * same state.  We can process that many blocks in one go.
@@ -316,6 +375,8 @@ blk_read_multiple (struct blk_overlay *blk,
         bitmap_get_blk (&blk->bm, blknum + b, BLOCK_NOT_ALLOCATED);
       if (state != s)
         break;
+      if (get_fd_for_blknum (blk, blknum + b) != fd)
+        break;
     }
   }
 
@@ -405,7 +466,7 @@ blk_cache (struct blk_overlay *blk,
   off_t offset = blknum * blksize;
   enum bm_entry state = bitmap_get_blk (&blk->bm, blknum, BLOCK_NOT_ALLOCATED);
   unsigned n = blksize, tail = 0;
-  const int fd = blk->fds.ptr[0];
+  const int fd = get_fd_for_blknum (blk, blknum);
 
   if (offset + n > blk->size) {
     tail = offset + n - blk->size;
@@ -458,7 +519,7 @@ blk_write (struct blk_overlay *blk,
            uint64_t blknum, const uint8_t *block, int *err)
 {
   off_t offset = blknum * blksize;
-  const int fd = blk->fds.ptr[0];
+  const int fd = get_fd_for_blknum (blk, blknum);
 
   if (cow_debug_verbose)
     nbdkit_debug ("cow: blk_write block %" PRIu64 " (offset %" PRIu64 ")",
diff --git a/tests/Makefile.am b/tests/Makefile.am
index d7053ba2..bc66b61c 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -1724,7 +1724,10 @@ TESTS += \
 	test-cow-unaligned.sh \
 	$(NULL)
 endif
-TESTS += test-cow-null.sh
+TESTS += \
+	test-cow-huge.sh \
+	test-cow-null.sh \
+	$(NULL)
 EXTRA_DIST += \
 	test-cow.sh \
 	test-cow-block-size.sh \
@@ -1732,6 +1735,7 @@ EXTRA_DIST += \
 	test-cow-extents1.sh \
 	test-cow-extents2.sh \
 	test-cow-extents-large.sh \
+	test-cow-huge.sh \
 	test-cow-null.sh \
 	test-cow-on-read.sh \
 	test-cow-on-read-caches.sh \
diff --git a/tests/test-cow-huge.sh b/tests/test-cow-huge.sh
new file mode 100755
index 00000000..cad1f6cc
--- /dev/null
+++ b/tests/test-cow-huge.sh
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
+# nbdkit
+# Copyright Red Hat
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+# * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution.
+#
+# * Neither the name of Red Hat nor the names of its contributors may be
+# used to endorse or promote products derived from this software without
+# specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY RED HAT AND CONTRIBUTORS ''AS IS'' AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+# PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL RED HAT OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+# USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+# OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+
+# Test cow filter with huge overlays.  This used to break because of
+# file size limits on /var/tmp.
+
+source ./functions.sh
+set -e
+set -x
+set -u
+
+requires_run
+requires_nbdsh_uri
+
+# ext4 has a 16T file size limit.
+size=17T
+
+define script <<'EOF'
+
+import random
+
+random.seed(None)
+
+M = 1024*1024
+T = 1024*1024*1024*1024
+
+# Test a few megabytes either side of the 8T boundary.
+overlay = bytearray(10 * M)
+overlay_offset = int(8 * T - len(overlay)/2)
+assert h.get_size() > overlay_offset + len(overlay)
+
+for iter in range(1, 200):
+    offset = random.randint(0, len(overlay) - 10)
+    count = min(len(overlay) - offset, random.randint(1, 1 * M))
+    buf = h.pread(count, overlay_offset + offset)
+    assert buf == overlay[offset:offset+count]
+    c = random.randint(1, 255)
+    buf = bytearray([c]) * count
+    h.pwrite(buf, overlay_offset + offset)
+    overlay[offset:offset+count] = buf
+    buf = h.pread(count, overlay_offset + offset)
+    assert buf == overlay[offset:offset+count]
+
+    if random.randint(0, 10) >= 7:
+        offset = random.randint(0, len(overlay) - 10)
+        count = min(len(overlay) - offset, random.randint(1, 1 * M))
+        zbuf = bytearray(count)
+        h.trim(count, overlay_offset + offset)
+        overlay[offset:offset+count] = zbuf
+        buf = h.pread(count, overlay_offset + offset)
+        assert buf == zbuf
+
+EOF
+export script
+
+nbdkit -fv --filter=cow null $size --run ' nbdsh -u "$uri" -c "$script" '
-- 
2.47.3

